localnodedns/controllers/localnodedns_controller.go

```diff
@@ -0,0 +1,117 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/addon"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/addon/pkg/status"
+	"sigs.k8s.io/kubebuilder-declarative-pattern/pkg/patterns/declarative"
+
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	api "sigs.k8s.io/cluster-addons/localnodedns/api/v1alpha1"
+)
+
+var _ reconcile.Reconciler = &LocalNodeDNSReconciler{}
+
+// LocalNodeDNSReconciler reconciles a LocalNodeDNS object
+type LocalNodeDNSReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	declarative.Reconciler
+	watchLabels declarative.LabelMaker
+}
+
+func (r *LocalNodeDNSReconciler) setupReconciler(mgr ctrl.Manager) error {
+	addon.Init()
+
+	labels := map[string]string{
+		"k8s-app": "localnodedns",
+	}
+
+	r.watchLabels = declarative.SourceLabel(mgr.GetScheme())
+
+	return r.Reconciler.Init(mgr, &api.LocalNodeDNS{},
+		declarative.WithRawManifestOperation(replaceVariables(mgr)),
+		declarative.WithObjectTransform(declarative.AddLabels(labels)),
+		declarative.WithOwner(declarative.SourceAsOwner),
+		declarative.WithLabels(r.watchLabels),
+		declarative.WithStatus(status.NewBasic(mgr.GetClient())),
+		declarative.WithObjectTransform(addon.TransformApplicationFromStatus),
+		declarative.WithManagedApplication(r.watchLabels),
+		declarative.WithObjectTransform(addon.ApplyPatches),
+	)
+}
+
+// +kubebuilder:rbac:groups=addons.x-k8s.io,resources=localnodedns,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=addons.x-k8s.io,resources=localnodedns/status,verbs=get;update;patch
+
+func (r *LocalNodeDNSReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	if err := r.setupReconciler(mgr); err != nil {
+		return err
+	}
+
+	c, err := controller.New("localnodedns-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to LocalNodeDNS
+	err = c.Watch(&source.Kind{Type: &api.LocalNodeDNS{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to deployed objects
+	_, err = declarative.WatchAll(mgr.GetConfig(), c, r, r.watchLabels)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func replaceVariables(mgr ctrl.Manager) declarative.ManifestOperation {
+	return func(ctx context.Context, object declarative.DeclarativeObject, s string) (string, error) {
+		o := object.(*api.LocalNodeDNS)
+		kubeProxyMode, err := findKubeProxyMode(ctx, mgr.GetClient())
+		if err != nil {
+			fmt.Println("Error determining kube-proxy mode: Defaulting to iptables")
```

Comment:
In general, use logging functions over stdout / stderr.  So either `klog.Warningf("error determining kube-proxy mode...`) or `log.Log("error determining kube-proxy mode...`.

You also should include the error, otherwise it's really hard for someone that sees the log to know what went wrong; for example:

`klog.Warningf("error determining kube-proxy mode, defaulting to iptables: %v", err)`

I think log has a dedicated helper:

`log.Error(err, "error determining kube-proxy mode, defaulting to iptables")` 

Logging in kubernetes is generally in flux right now; we're moving to structured logging (log) but doing so by adding it to klog (I think)  https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1602-structured-logging